lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -169,6 +221,79 @@ static LogicalResult verifyComponentOp(ComponentOp op) {
             "\"calyx.cells\", \"calyx.wires\", \"calyx.control\".";
 }
 
+//===----------------------------------------------------------------------===//
+// CellsOp
+//===----------------------------------------------------------------------===//
+static LogicalResult verifyCellsOp(CellsOp cells) {
+  auto component = cells->getParentOfType<ComponentOp>();
+  if (!component)
+    return cells.emitOpError("should be embedded in 'calyx.component'.");
+
+  if (!llvm::all_of(*cells.getBody(), [](auto &op) { return isa<CellOp>(op); }))
+    return cells.emitOpError("should only contain 'calyx.cell' ops.");
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// CellOp
+//===----------------------------------------------------------------------===//
+
+/// Lookup the component for the symbol. This returns null on
+/// invalid IR.
+Operation *CellOp::getReferencedComponent() {
+  auto program = (*this)->getParentOfType<ProgramOp>();
+  if (!program)
```

Comment:
I think this is guaranteed by the `HasParent` trait (that trait should insert this exact check into the verifier). It seems like you don't need to defensively program around that case here.